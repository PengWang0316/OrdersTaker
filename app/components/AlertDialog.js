import React from 'react';
import PropTypes from 'prop-types';
import { Dialog, DialogTitle, DialogContent, DialogContentText, DialogActions, Button } from '@material-ui/core';

export const AlertDialog = ({
  open, onClose, title, content, onSecondButton, onFirstButton, firstButtonText, secondButtonText
}) => (
  <Dialog
    open={open}
    onClose={onClose}
    aria-labelledby="alert-dialog-title"
    aria-describedby="alert-dialog-description"
  >
    <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
    <DialogContent>
      <DialogContentText id="alert-dialog-description">{content}</DialogContentText>
    </DialogContent>
    <DialogActions>
      <Button onClick={onFirstButton} color="primary">
        {firstButtonText}
      </Button>
      <Button onClick={onSecondButton} color="primary" autoFocus>
        {secondButtonText}
      </Button>
    </DialogActions>
  </Dialog>
);
AlertDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  onFirstButton: PropTypes.func.isRequired,
  onSecondButton: PropTypes.func.isRequired,
  firstButtonText: PropTypes.string,
  secondButtonText: PropTypes.string
};
AlertDialog.defaultProps = {
  firstButtonText: 'Cancel',
  secondButtonText: 'Confirm'
};
export default AlertDialog;
