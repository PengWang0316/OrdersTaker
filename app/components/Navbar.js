import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  MenuItem, Menu, IconButton, Hidden, Button, Typography, Toolbar, AppBar, Avatar
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import MenuIcon from '@material-ui/icons/Menu';

// import LoginDialog from './LoginDialog/';
// import LoginDialogSnackbar from './snackbars/LoginDialogSnackbar';
// import LogoutSnackbar from './snackbars/LogoutSnackbar';
import { logout as logoutActon } from '../actions/UserActions';
import {
  LOGIN_CALLBACK_URL, HOME_PAGE_URL, ORDERS_PAGE_URL, NORMAL_USER_ROLE, KITHEN_INTERFACE_PAGE_URL
} from '../config';
import LoginDialogContext from '../contexts/LoginDialogContext';

const URL_REGEXP = /^https?:\/\/.+?(\/.*)/; // Using this to get the relative url.

/* istanbul ignore next */
const styles = theme => ({
  link: {
    color: theme.palette.primary.contrastText,
    textDecoration: 'none'
  },
  menuLink: {
    textDecoration: 'none'
  },
  flex1: {
    flex: 1
  },
  appbar: {
    maxHeight: 55,
    position: 'sticky',
    top: 0
  },
  avatar: {
    width: 30,
    height: 30,
    marginRight: 8
  }
});

/** Navbar component */
export class Navbar extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
    logout: PropTypes.func.isRequired
  };

  state = {
    anchorEl: null // The indecator for the menu
    // open: false, // The indecator for the loginDialog,
    // snackbarOpen: false, // For the warning snackbar.
    // snackbarMessage: '',
    // logoutSnackbarOpen: false
  };

  /**
   * Changing anchorEl state to an click target element or null.
   * @return {null} No Return.
   */
  handleMenuIconClick = ({ currentTarget }) => this.setState(({ anchorEl }) => {
    if (anchorEl) return { anchorEl: null };
    return { anchorEl: currentTarget };
  });

  /**
   * Showing the login dialog when the user did not login and logout a user if the user has already login.
   * @return {null} No return.
   */
  handleLoginButtonClick = () => {
    const { user, logout, history } = this.props;
    if (user._id) { // If logout button was clicked, clear the state open in order to prevent the LoginDialog shows up.
      // this.setState({ open: false, logoutSnackbarOpen: true });
      this.handleLogoutAction();
      logout();
      history.push(HOME_PAGE_URL);
    } else {
      this.setState({ anchorEl: null }); // Have to make sure always close the menu.
      const matchUrl = window.location.href.match(URL_REGEXP);
      /* istanbul ignore next */
      localStorage.setItem(LOGIN_CALLBACK_URL, matchUrl ? matchUrl[1] : '/'); // Save the current url to the local storage, which can be used to redirect users to the page they are useing after login.
      // this.handleToggleDialog();
      this.handleToggleLoginDialog();
    }
  };

  /**
   * The render method to render the jsx.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes, user } = this.props;
    const { anchorEl } = this.state;
    return (
      <Fragment>
        <LoginDialogContext.Consumer>
          {({ handleLogoutAction, handleToggleLoginDialog }) => {
            this.handleLogoutAction = handleLogoutAction;
            this.handleToggleLoginDialog = handleToggleLoginDialog;
            return (
              <AppBar position="static" className={classes.appbar} data-testid="navbar">
                <Toolbar>
                  <Link to={HOME_PAGE_URL} className={`${classes.link} ${classes.flex1}`} data-testid="titleLink">
                    <Typography variant="title" color="inherit">Name of the restaurant</Typography>
                  </Link>
                  <Hidden only="xs">
                    <Link to={ORDERS_PAGE_URL} className={classes.link}>
                      <Button color="inherit" data-testid="orderButton">Order</Button>
                    </Link>
                    <Button color="inherit" data-testid="menuButton">Menu</Button>
                    {user.role && user.role < NORMAL_USER_ROLE && (
                      <Link to={KITHEN_INTERFACE_PAGE_URL} className={classes.link}>
                        <Button color="inherit" data-testid="kithenButton">Kithen</Button>
                      </Link>
                    )}
                    <Button color="inherit" onClick={this.handleLoginButtonClick}  data-testid="loginButton">
                      {user._id ? (
                        <Fragment>
                          {user.avatar && <Avatar alt="avatar" className={classes.avatar} src={user.avatar} />}
                          <Typography color="inherit">Logout</Typography>
                        </Fragment>) : 'Login'}
                    </Button>
                  </Hidden>
                  <Hidden only={['xl', 'lg', 'md', 'sm']}>
                    <IconButton
                      color="inherit"
                      aria-label="Menu"
                      onClick={this.handleMenuIconClick}
                      aria-owns={anchorEl ? 'simple-menu' : null}
                      aria-haspopup="true"
                      data-testid="navbarDropMenuButton"
                    >
                      <MenuIcon />
                    </IconButton>
                    <Menu
                      id="simple-menu"
                      anchorEl={anchorEl}
                      open={Boolean(anchorEl)}
                      onClose={this.handleMenuIconClick}
                      data-testid="dropDownMenu"
                    >
                      <MenuItem>
                        <Link to={ORDERS_PAGE_URL} className={classes.menuLink} data-testid="orderLink">
                          <Typography color="textPrimary">Order</Typography>
                        </Link>
                      </MenuItem>
                      <MenuItem>
                        <Typography color="textPrimary">Menu</Typography>
                      </MenuItem>
                      {user.role && user.role < NORMAL_USER_ROLE && (
                        <MenuItem>
                          <Link to={KITHEN_INTERFACE_PAGE_URL} className={classes.menuLink} data-testid="kithenLink">
                            <Typography color="textPrimary">Kithen</Typography>
                          </Link>
                        </MenuItem>
                      )}
                      <MenuItem onClick={this.handleLoginButtonClick} data-testid="loginMenu">
                        {user._id ? (
                          <Fragment>
                            {user.avatar && <Avatar alt="avatar" className={classes.avatar} src={user.avatar} />}
                            <Typography color="textPrimary">Logout</Typography>
                          </Fragment>) : 'Login'}
                      </MenuItem>
                    </Menu>
                  </Hidden>
                </Toolbar>
              </AppBar>
            );
          }}
        </LoginDialogContext.Consumer>
      </Fragment>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logoutActon())
});
// Putting the withRouter to the first position because when test code mocks Link, the withRouter also has to be mocked. But it is hard to really return a react component to satisfy the whole chain call.
export default withRouter(connect(mapStateToProps, mapDispatchToProps)((withStyles(styles)(Navbar))));
