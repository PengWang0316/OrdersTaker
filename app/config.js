export const JWT_MESSAGE = 'OrdersTaker_JWT';
export const ADMINISTRATOR_ROLE = 1;
export const SUPER_USER_ROLE = 2;
export const NORMAL_USER_ROLE = 3;
// export const BASE_URL = 'https://orders-taker.glitch.me'; // Development server
// export const BASE_URL = 'https://orderstaker.kevin-project.com:8081'; // Development server on the AWS
export const BASE_URL = 'https://orderstaker.kevin-project.com:8080'; // Production server
// export const SOCKETIO_URL = 'https://orderstaker.kevin-project.com:3001';
export const SOCKETIO_URL = 'https://orderstaker.kevin-project.com:3000'; // Production server

export const CLOUDINARY_UPLOAD_PRESET = 'OrdersTakerWebApp'; // The preset of Cloudinary.
export const LOGIN_CALLBACK_URL = 'OrdersTaker_loginCallbackUrl';

export const ITEM_ONE_PRICE_KEY = '_onePrice';
export const MAX_ORDER_AMOUNT = 6; // How many orders will show in the one page
export const ORDER_STATUS_FINISHED = 'Finished';
export const ORDER_STATUS_RECEIVED = 'Received';
export const ORDER_STATUS_PREPARING = 'Preparing';

/* Lazy image loading */
export const LAZY_IMAGE_CLASS = 'lazyImage';
export const IMAGE_PLACEHOLDER_URL = 'https://res.cloudinary.com/orderstaker/image/upload/v1536616770/others/PlaceHolder.png';

/* URLS */
export const HOME_PAGE_URL = '/';
export const LOGIN_REDIRECT_RUL = '/loginRedirect';
export const CART_PAGE_URL = '/cartPage';
export const ORDERS_PAGE_URL = '/ordersPage';
export const KITHEN_INTERFACE_PAGE_URL = '/kithenInterfacePage';

/* Socket IO events */
export const SOCKETIO_EVENT_ADD_NEW_ORDER = 'socketioEventAddNewOrder';
export const SOCKETIO_EVENT_UPDATE_ORDER_ITEM = 'socketioEventUpdateOrderItem';
