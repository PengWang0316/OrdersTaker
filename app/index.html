<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="A simply order taking system for restaurants">
    <meta name="keywords" content="Order taker restaurant free">
    <!-- Add manifest -->
    <link rel="manifest" href="/manifest.json">
    <!-- Tell the browser it's a PWA -->
    <meta name="mobile-web-app-capable" content="yes">
    <!-- Tell iOS it's a PWA -->
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="theme-color" content="#6d9cb5"/>
    <link rel="icon" href="/favicon.png" >
    <!-- Latest compiled and minified CSS -->

    <title>Orders Taker - A simply order taking system for restaurants</title>
  </head>
  <body>
    <div id="app"></div>
    <noscript>This site requires JavaScript to work. Please change to a browser that supports JavaScript or turn on JavaScript in the setting.</noscript>
    <!-- Latest compiled and minified JavaScript -->
  </body>
</html>
