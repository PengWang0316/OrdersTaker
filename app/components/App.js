
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
// import Loadable from 'react-loadable'; Does not work with react-hot-loader very well.
// import universal from 'react-universal-component'; Does not work with react-hot-loader very well.
import importedComponent from 'react-imported-component';
import CssBaseline from '@material-ui/core/CssBaseline';

import LoginDialog from './LoginDialog/';
import LoginDialogSnackbar from './snackbars/LoginDialogSnackbar';
import LogoutSnackbar from './snackbars/LogoutSnackbar';
import Navbar from './Navbar';
import { parserUserFromJwt } from '../actions/UserActions';
import LoginDialogContext from '../contexts/LoginDialogContext';

import { JWT_MESSAGE, HOME_PAGE_URL, LOGIN_REDIRECT_RUL, ORDER_PAGE_URL, ORDER_STATUS_PAGE_URL, ORDERS_PAGE_URL } from '../config';
import LoadingAnimation from './SharedComponents/LoadingAnimation';

// import { CSSTransitionGroup } from 'react-transition-group';

// import BackgroundMainPager from './BackgroundMainPager';

const theme = createMuiTheme({
  palette: {
    primary: {
      light: '#9ecde7',
      main: '#6d9cb5',
      dark: '#3e6e85',
      contrastText: '#fff',
    },
    secondary: {
      light: '#497ca4',
      main: '#105075',
      dark: '#002884',
      contrastText: '#fff',
    },
  },
});

/* istanbul ignore next */
const HomePage = importedComponent(() => import(/* webpackChunkName: "HomePageContainer" */ './containers/HomePageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const LoginRedirect = importedComponent(() => import(/* webpackChunkName: "LoginRedirect" */ './LoginRedirect').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderPage = importedComponent(() => import(/* webpackChunkName: "OrderPageContainer" */ './containers/OrderPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrderStatusPage = importedComponent(() => import(/* webpackChunkName: "OrderStatusPageContainer" */ './containers/OrderStatusPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/* istanbul ignore next */
const OrdersPage = importedComponent(() => import(/* webpackChunkName: "OrdersPageContainer" */ './containers/OrdersPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/**
 * The root component that contains the theme, routers, navbar, and login dialog
 */
export class App extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    parserUserFromJwt: PropTypes.func.isRequired
  };

  /**
   * If Redux does not have user and local storage has jwt, use jwt to parser the user.
   * @param {object} props contains the component's prop value.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (!props.user._id) {
      const jwtMessage = localStorage.getItem(JWT_MESSAGE);
      if (jwtMessage) props.parserUserFromJwt(jwtMessage);
    }
  }

  state = {
    isLoginDialogOpen: false,
    isLogoutSnackBarOpen: false,
    isLoginSnackbarOpen: false,
    snackbarMessage: ''
  };

  /**
   * Setting the open state to an opposite value.
   * @return {null} No return.
   */
  handleToggleDialog = () => this.setState(({ isLoginDialogOpen }) => ({ isLoginDialogOpen: !isLoginDialogOpen }));

  /**
   * Setting the snackbarOpen state to true in order to open the snackbar.
   * If message paramter is a string, set it to message state. Otherwise, set a empty message.
   * @param {string} message will be showed in the snackbar.
   * @return {null} No return.
   */
  handleToggleSnackbar = message => this.setState(({ isLoginSnackbarOpen }) =>
    ({ isLoginSnackbarOpen: !isLoginSnackbarOpen, snackbarMessage: typeof message === 'string' ? message : '' }));

  /**
   * Set the logoutSnackbarOpen state to an opposite value.
   * @return {null} No return.
   */
  handleToggleLogoutSnackbar = () => this.setState(({ isLogoutSnackBarOpen }) =>
    ({ isLogoutSnackBarOpen: !isLogoutSnackBarOpen }));

  /**
   * It will be pass down in a context in order to make it available for all component.
   * This method will be used when logout action is performed.
   * Clear the state isLoginDialogOpen in order to prevent the LoginDialog shows up.
   * @return {null} No return.
   */
  handleLogoutAction = () => this.setState({ isLoginDialogOpen: false, isLogoutSnackBarOpen: true });

  /**
   * The render method for the component
   * @return {jsx} Return jsx
   */
  render() {
    const { user } = this.props;
    const { isLogoutSnackBarOpen, isLoginDialogOpen, isLoginSnackbarOpen, snackbarMessage } = this.state;
    return (
      <MuiThemeProvider theme={theme}>
        <Router>
          {/* Using a context to give children components access to login dialog */}
          <LoginDialogContext.Provider value={{ handleLogoutAction: this.handleLogoutAction, handleToggleLoginDialog: this.handleToggleDialog }}>
            <div>
              <CssBaseline />
              <Navbar />
              <main>
                <Switch>
                  <Route exact path={HOME_PAGE_URL} component={HomePage} />
                  <Route exact path={ORDER_PAGE_URL} component={OrderPage} />
                  <Route exact path={LOGIN_REDIRECT_RUL} component={LoginRedirect} />
                  <Route path={`${ORDER_STATUS_PAGE_URL}/:orderId`} component={OrderStatusPage} />
                  <Route path={ORDERS_PAGE_URL} component={OrdersPage} />
                  <Route render={() => <p>Not Fount!</p>} />
                </Switch>
              </main>
              {!user._id && ( // If a user has already login, do not need to load componnets below.
                <Fragment>
                  <LoginDialog open={isLoginDialogOpen} onClose={this.handleToggleDialog} onToggleSnackbar={this.handleToggleSnackbar} />
                  <LoginDialogSnackbar open={isLoginSnackbarOpen} onClose={this.handleToggleSnackbar} message={snackbarMessage} />
                </Fragment>
              )}
              <LogoutSnackbar open={isLogoutSnackBarOpen} onClose={this.handleToggleLogoutSnackbar} />
            </div>
          </LoginDialogContext.Provider>
        </Router>
      </MuiThemeProvider>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  parserUserFromJwt: jwtMessage => dispatch(parserUserFromJwt(jwtMessage))
});
export default connect(mapStateToProps, mapDispatchToProps)(App);
