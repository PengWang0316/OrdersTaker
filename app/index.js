import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { AppContainer } from 'react-hot-loader';
import { PersistGate } from 'redux-persist/integration/react';

import App from './components/App';
import configureStore from './store/ConfigureStore';
import './styles/index.global.css';
import registerServiceWorker from './registerServiceWorker';

const { store, persistor } = configureStore();

// ReactDOM.render(
//   <Provider store={store}>
//     <App />
//   </Provider>,
//   document.getElementById('app')
// );
const render = Component => {
  ReactDOM.render(
    <AppContainer>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Component />
        </PersistGate>
      </Provider>
    </AppContainer>,
    document.getElementById('app'),
  );
};
render(App);

// Webpack Hot Module Replacement API
if (module.hot) module.hot.accept('./components/App', () => render(App));

// Registering a service worker in the production enviroment.
registerServiceWorker();
