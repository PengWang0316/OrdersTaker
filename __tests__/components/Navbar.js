import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import { Navbar } from '../../app/components/Navbar';
import { LOGIN_CALLBACK_URL, HOME_PAGE_URL } from '../../app/config';
import context from '../../app/contexts/LoginDialogContextTestHelper';

jest.mock('@material-ui/core/AppBar', () => 'AppBar');
jest.mock('@material-ui/core/Toolbar', () => 'Toolbar');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Hidden', () => 'Hidden');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/Menu', () => 'Menu');
jest.mock('@material-ui/core/MenuItem', () => 'MenuItem');
jest.mock('@material-ui/core/Avatar', () => 'Avatar');
jest.mock('@material-ui/icons/Menu', () => 'MenuIcon');
// jest.mock('react-router-dom/Link', () => 'Link');
jest.mock('react-router-dom', () => ({ Link: 'Link', withRouter: jest.fn() }));
// jest.mock('react-redux', () => ({ connect: jest.fn() }));
// jest.mock('@material-ui/core/styles', () => ({ withStyles: jest.fn() }));
jest.mock('../../app/contexts/LoginDialogContext'); // The __mocks__/LoginDialogContext.js will be used automatically.

// jest.mock('../../app/components/LoginDialog/LoginDialog', () => 'LoginDialog');
// jest.mock('../../app/components/snackbars/LoginDialogSnackbar', () => 'LoginDialogSnackbar');
// jest.mock('../../app/components/snackbars/LogoutSnackbar', () => 'LogoutSnackbar');
// jest.mock('@material-ui/core/styles', () => { withStyles });

describe('Navbar test', () => {
  const defaultProps = {
    classes: {
      link: 'link', appbar: 'appbar', menuLink: 'menuLink', flex1: 'flex1', avatar: 'avatar'
    },
    logout: jest.fn(),
    user: {},
    history: { push: jest.fn() }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<Navbar {...props} />);

  test('Initial state and parserUserFromJwt', () => {
    const component = getShallowComponent();
    expect(component.state('anchorEl')).toBe(null);
  });

  test('handleMenuIconClick', () => {
    const component = getShallowComponent();
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toEqual('currentTarget');
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toBe(null);
  });

  test('handleLoginButtonClick', () => {
    window.console = { // Silence the error and warning that come from enzyme mount.
      error: jest.fn(),
      warning: jest.fn()
    };
    const component = mount(<Navbar {...defaultProps} />);
    component.setState({ anchorEl: true });
    component.instance().handleLoginButtonClick();
    expect(defaultProps.logout).not.toHaveBeenCalled();
    expect(component.state('anchorEl')).toBeNull();
    // expect(localStorage.setItem).toHaveBeenCalledTimes(1);
    // expect(localStorage.setItem).toHaveBeenLastCalledWith(LOGIN_CALLBACK_URL, '/');
    expect(localStorage.getItem(LOGIN_CALLBACK_URL)).toEqual('/');
    expect(context.handleLogoutAction).not.toHaveBeenCalled();
    expect(context.handleToggleLoginDialog).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).not.toHaveBeenCalled();

    // global.window = { location: { href: { value: 'http://df:39/dd' } } };
    // Object.defineProperty(window.location, 'href', {
    //   writable: true,
    //   value: 'http://df:39/dd'
    // });
    jsdom.reconfigure({ // changing the window.location.href
      url: 'http://df:39/dd'
    });
    component.instance().handleLoginButtonClick();
    // expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    // expect(localStorage.setItem).toHaveBeenLastCalledWith(LOGIN_CALLBACK_URL, '/dd');
    expect(localStorage.getItem(LOGIN_CALLBACK_URL)).toEqual('/dd');
    component.setProps({ user: { _id: 'id' } }); // Setting a user object to props in order to test handleLoginButtonClick function.
    component.instance().handleLoginButtonClick();
    expect(defaultProps.logout).toHaveBeenCalledTimes(1);
    expect(context.handleLogoutAction).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenLastCalledWith(HOME_PAGE_URL);
    // expect(component.state('open')).toBe(false);
  });

  test('NavBar snapshot without user', () => expect(renderer.create(<Navbar {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('NavBar snapshot with user without avatar', () => expect(renderer.create(<Navbar {...{ ...defaultProps, user: { _id: 'id' } }} />).toJSON()).toMatchSnapshot());

  test('NavBar snapshot with user with avatar', () => expect(renderer.create(<Navbar {...{ ...defaultProps, user: { _id: 'id', avatar: 'avatar' } }} />).toJSON()).toMatchSnapshot());

  test('NavBar snapshot with user with avatar with role 2', () => expect(renderer.create(<Navbar {...{ ...defaultProps, user: { _id: 'id', avatar: 'avatar', role: 2 } }} />).toJSON()).toMatchSnapshot());
});
