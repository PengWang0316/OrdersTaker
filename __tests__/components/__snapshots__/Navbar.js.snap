// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar test NavBar snapshot with user with avatar 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="title"
      >
        Name of the restaurant
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/ordersPage"
      >
        <Button
          color="inherit"
          data-testid="orderButton"
        >
          Order
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="menuButton"
      >
        Menu
      </Button>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        <Avatar
          alt="avatar"
          className="avatar"
          src="avatar"
        />
        <Typography
          color="inherit"
        >
          Logout
        </Typography>
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="orderLink"
            to="/ordersPage"
          >
            <Typography
              color="textPrimary"
            >
              Order
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            Menu
          </Typography>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          <Avatar
            alt="avatar"
            className="avatar"
            src="avatar"
          />
          <Typography
            color="textPrimary"
          >
            Logout
          </Typography>
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;

exports[`Navbar test NavBar snapshot with user with avatar with role 2 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="title"
      >
        Name of the restaurant
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/ordersPage"
      >
        <Button
          color="inherit"
          data-testid="orderButton"
        >
          Order
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="menuButton"
      >
        Menu
      </Button>
      <Link
        className="link"
        to="/kithenInterfacePage"
      >
        <Button
          color="inherit"
          data-testid="kithenButton"
        >
          Kithen
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        <Avatar
          alt="avatar"
          className="avatar"
          src="avatar"
        />
        <Typography
          color="inherit"
        >
          Logout
        </Typography>
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="orderLink"
            to="/ordersPage"
          >
            <Typography
              color="textPrimary"
            >
              Order
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            Menu
          </Typography>
        </MenuItem>
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="kithenLink"
            to="/kithenInterfacePage"
          >
            <Typography
              color="textPrimary"
            >
              Kithen
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          <Avatar
            alt="avatar"
            className="avatar"
            src="avatar"
          />
          <Typography
            color="textPrimary"
          >
            Logout
          </Typography>
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;

exports[`Navbar test NavBar snapshot with user without avatar 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="title"
      >
        Name of the restaurant
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/ordersPage"
      >
        <Button
          color="inherit"
          data-testid="orderButton"
        >
          Order
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="menuButton"
      >
        Menu
      </Button>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        <Typography
          color="inherit"
        >
          Logout
        </Typography>
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="orderLink"
            to="/ordersPage"
          >
            <Typography
              color="textPrimary"
            >
              Order
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            Menu
          </Typography>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          <Typography
            color="textPrimary"
          >
            Logout
          </Typography>
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;

exports[`Navbar test NavBar snapshot without user 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="title"
      >
        Name of the restaurant
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/ordersPage"
      >
        <Button
          color="inherit"
          data-testid="orderButton"
        >
          Order
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="menuButton"
      >
        Menu
      </Button>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        Login
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="orderLink"
            to="/ordersPage"
          >
            <Typography
              color="textPrimary"
            >
              Order
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            Menu
          </Typography>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          Login
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;
